Add hive-consensus-tests workflow and program to generate it

marcindsobczak opened the pull request:
## Changes

This PR introduces
- optimized workflow file, which is splitting hive consensus tests to 53 concurrent jobs
- simple generator which is preparing workflow above and can be used to update workflow in the future, when new tests would be added

### How generator works?
It is splitting tests into jobs basing on size of test files. It's not an optimal way (sometimes big files are quick to test), but there are no small files which are tested long time, so using that logic we were able to create jobs not-exceeding hard 6h github timeout.
It is reading all test directories. If directory is smaller than 35MB, would be testes as a whole. For bigger directories, each .json file is splitted.
Directories/single .json tests bigger than 23MB are tested in dedicated job. Smaller directories/single tests are combined into jobs with max 23MB.
With current shape of ethereum tests, we are splitting tests to 53 jobs. If more tests would be added, generator would create more jobs. We should always be able to finish workflow below 6h.
There are a few single .json files (inseparable) which are running ~5h, so creating more jobs was pointless, because we are not able to finish workflow below 5h anyway.

### Input value: parallelism
Hive tests (running inside jobs) can be run in parallel. Without concurrency we are not able to finish workflow in time - testing biggest single .json files is taking longer than 6h. With parallelism of 2 we are sometimes finishing successfully and sometimes hitting timeout as well. On beefy machines it is possible to run even 16 tests at once. Unfortunately, github infrastructure is not allowing us to go that far - we have false-positive fails with even 4 concurrently running tests. We need to run parallelism of 3 (3 is a bit surprising number for me, but is working pretty well - no increased flakiness and no timeouts, looks like sweat spot)
Parallelism is an input value with possible options of 1, 2, 3, 4, 8, 16 and default value of 3.

### Test runs
With parallelism of 3 we had one perfect run (failing only tests of EOF, like on official hive):
https://github.com/NethermindEth/nethermind/actions/runs/4227312341
And 2 runs with some false-positives:
https://github.com/NethermindEth/nethermind/actions/runs/4226285955
https://github.com/NethermindEth/nethermind/actions/runs/4228090797

What is a good news, total duration is repeatable and is 5h10min-5h20min. What is a bad news, we had only one run without false-positives, and two with 1 (in 1 job) and 4 (in 3 jobs) false-positives, which mean a need of re-run accordingly 1 and 3 jobs and wait another 1,5h-4,5h for results (or run failing tests as single tests in general `hive-tests.yml` with limit and have it in ~20min).

Anyway, without this workflow we needed to run it on external machines and wait >30h, so running it in ~5h15min and even potentially rerunning some jobs is a big step forward.

Current design, after last-minute cosmetic changes:
https://github.com/NethermindEth/nethermind/actions/runs/4243023659

I already worked on creating a workflow to automatically run generator and update `hive-consensus-tests.yml`, but I ran into some difficulties and decided to add it in another PR in the future to not block functionalities introduced in this PR.

## Types of changes

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

#### Requires testing

- [ ] Yes
- [x] No

Tested manually

LukaszRozmej commented on src/Nethermind/Nethermind.Hive.ConsensusWorkflowGenerator/Program.cs:
```diff
@@ -0,0 +1,224 @@
+
+    private static List<List<string>> GetTestsSplittedToJobs(Dictionary<string, long> pathsToBeTested)
+    {
+        List<List<string>> accumulatedJobs = new();
```
small - would make it lazy enumerable :)

LukaszRozmej commented on src/Nethermind/Nethermind.Hive.ConsensusWorkflowGenerator/Program.cs:
```diff
@@ -0,0 +1,224 @@
+    {
+        string testsDirectory = string.Concat(FindDirectory("tests"), "/BlockchainTests");
+
+        List<string> directories = Directory.GetDirectories(testsDirectory, "st*", SearchOption.AllDirectories).ToList();
+        directories.AddRange(Directory.GetDirectories(testsDirectory, "bc*", SearchOption.AllDirectories).ToList());
+
+        return directories;
+    }
```
could be lazy enumerated instaed of list

LukaszRozmej commented on src/Nethermind/Nethermind.Hive.ConsensusWorkflowGenerator/Program.cs:
```diff
@@ -0,0 +1,224 @@
+    {
+        StringBuilder fileContent = new();
+
+        List<string> directories = GetTestsDirectories();
```
Would make sense to take some parameter from command line about the path to tests

LukaszRozmej commented on src/Nethermind/Nethermind.Hive.ConsensusWorkflowGenerator/Program.cs:
```diff
@@ -0,0 +1,224 @@
+        File.WriteAllText($"{FindDirectory("Nethermind.Hive.ConsensusWorkflowGenerator")}/hive-consensus-tests.yml", fileContent.ToString());
+    }
+
+    private static Dictionary<string, long> GetPathsToBeTested(List<string> directories)
```
Could just return lazy IEnumerable<KeyValuePair>

marcindsobczak replied:
In this place I can't. I need to have access to already added paths when adding new one to avoid double-testing some paths and exceeding assumed max size of single job. If there are tests with the same name in different directories e.g. `exampleTestName.json` in both `stExampleA` and `stExampleB` directories, hive `--sim.limit /exampleTestName` will test `jsons` from both directories. Because of that I want to add it only once and then if `exampleTestName` would be already present in dictionary, only sum sizes of already existed and newly added. And because of that I can't `yield return` after first occurrence

LukaszRozmej commented on src/Nethermind/Nethermind.Hive.ConsensusWorkflowGenerator/Program.cs:
```diff
@@ -0,0 +1,224 @@
+        int jobsCreated = 0;
+        foreach (List<string> job in accumulatedJobs)
+        {
+            WriteJob(fileContent, job, ++jobsCreated);
```
Better to write directly to file, which should be opened here and passed as TextWriter, instead of to StringBuilder

LukaszRozmej commented on .github/workflows/hive-consensus-tests.yml:
```diff
@@ -0,0 +1,4522 @@
+        uses: actions/checkout@v3
+        with:
+          path: nethermind
+      - name: Set up QEMU
```
not sure if something can be cached: https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows

marcindsobczak replied:
Not sure too, I will left it as it is, might be potentially optimized in the future

LukaszRozmej commented:
I think the generation should be moved to tools directory, the output of the generation should by default modify the workflow file.

Which input should it grab by default? Our tests subrepo?

MarekM25 commented:
> I think the generation should be moved to tools directory, the output of the generation should by default modify the workflow file.
> 
> Which input should it grab by default? Our tests subrepo?

I agree with Łukasz, let's use tools directory

marcindsobczak commented:
Made some changes:
- moved project to tools
- `hive-consensus-tests.yml` is saved directly in final destination - `nethermind/.github/workflows`
- path can be passed as program argument. If there is no argument, default is `src/tests` so directory of `ethereum/tests` repo reference in our codebase
- `StringBuilder` is replaced by `TextWriter` and is writing directly to file
- introduced lazy enumeration where possible
